paddle/fluid/operators/center_loss_op.h

```diff
@@ -0,0 +1,155 @@
+    auto *update_rate = ctx.Input<Tensor>("CenterUpdateRate");
+    int cluster_num = ctx.Attr<int>("cluster_num");
+    auto *lr_center = update_rate->data<T>();
+    T alpha = lr_center[0];
+    bool need_update = static_cast<T>(ctx.Attr<bool>("need_update"));
+
+    auto x_data = X->data<T>();
+    auto label_data = labels->data<int64_t>();
+
+    auto centers_dim = centers->dims();
+    auto centers_data = centers->data<T>();
+
+    auto x_dims = X->dims();
+    int batch_size = x_dims[0];
+    int deep_feat_dim = x_dims[1];
+
+    auto centers_diff = ctx.Output<Tensor>("CentersDiff");
+    auto centers_diff_data = centers_diff->mutable_data<T>(ctx.GetPlace());
+    auto *out_loss = ctx.Output<Tensor>("Loss");
+
+    auto *centers_out = ctx.Output<Tensor>("CentersOut");
+    auto *centers_out_data = centers_out->mutable_data<T>(ctx.GetPlace());
+
+    if (centers_out_data != centers_data) {
+      int size = centers_out->numel() * sizeof(T);
+      memcpy(centers_out_data, centers_data, size);
+    }
+
+    std::vector<int> center_update_count(cluster_num, 1);
+    auto &dev_ctx = ctx.template device_context<DeviceContext>();
+
+    auto loss_data = out_loss->mutable_data<T>(ctx.GetPlace());
+
+    Tensor centers_diffacc;  // used to accumulate all diff
+    auto centers_diffacc_data =
+        centers_diffacc.mutable_data<T>(centers_dim, ctx.GetPlace());
+    int numel = centers_diffacc.numel();
+    std::memset(centers_diffacc_data, 0, sizeof(T) * numel);
+
+    auto blas = math::GetBlas<DeviceContext, T>(ctx);
+    int tLabel;
+
+    T *x_index;
+    T *center_index;
+    T *center_loss_diff_index;
+    T *acc_index;
+    platform::Transform<DeviceContext> trans;
+
+    for (int i = 0; i < batch_size; ++i) {
+      tLabel = label_data[i];
+      center_update_count[tLabel]++;
+      x_index = const_cast<T *>(x_data) + i * deep_feat_dim;  // xi index
+      center_index = const_cast<T *>(centers_data) +
+                     tLabel * deep_feat_dim;  // center index
+      center_loss_diff_index = centers_diff_data + i * deep_feat_dim;
+      trans(dev_ctx, x_index, x_index + deep_feat_dim, center_index,
+            center_loss_diff_index, SubFunctor<T>());
+
+      acc_index = centers_diffacc_data + tLabel * deep_feat_dim;
+      blas.VADD(deep_feat_dim, center_loss_diff_index, acc_index,
+                acc_index);  // accumulate
+      loss_data[i] = blas.DOT(deep_feat_dim, center_loss_diff_index,
+                              center_loss_diff_index) /
+                     T(2.0);
+    }
+
+    // update centers data
+    if (need_update == true) {
+      for (int i = 0; i < cluster_num; i++) {
+        acc_index = centers_diffacc_data + i * deep_feat_dim;
+        center_index = centers_out_data + i * deep_feat_dim;
+        T scale = alpha / center_update_count[i];
+        blas.SCAL(deep_feat_dim, scale, acc_index);
+        blas.VADD(deep_feat_dim, acc_index, center_index, center_index);
```

Comment:
why does not CUDA kernel follow this calculation strategy?

Reply:
done.